### PR TITLE
have generated index and foreign key names match mysql

### PR DESCRIPTION
**Pull request by jycor**
changes:
- auto-generated secondary index names don't match (after a conflict)
  - we started at 0, while mysql starts at 2
- auto-generate foreign key names in a way that matches mysql
  - in-memory used to just have empty string as the constraint name

we used to error when generating foreign keys over the same sets of columns, but not anymore

related: https://github.com/dolthub/dolt/issues/7650

### CODE REVIEW

**Review by fulghum — APPROVED** [review]
Looks good! Nice work on the test cases and reverse engineering MySQL's behavior. Very thorough work there! 👏

File: enginetest/queries/foreign_key_queries.go [hunk 1]
```diff
@@ -2228,6 +2228,188 @@ var ForeignKeyTests = []ScriptTest{
...
+					{"child2", "CREATE TABLE `child2` (\n" +
+						"  `j` int NOT NULL,\n" +
+						"  PRIMARY KEY (`j`),\n" +
+						"  CONSTRAINT `child2_ibfk_1` FOREIGN KEY (`j`) REFERENCES `theparent` (`i`),\n" +
+						"  CONSTRAINT `child2_ibfk_2` FOREIGN KEY (`j`) REFERENCES `theparent` (`i`)\n" +
+						") ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_bin"},
+				},
+			},
+
+			// For some reason, the generated name will always be the highest one
```
Comment (zachmu) [inline enginetest/queries/foreign_key_queries.go:2299]: This is only the case when an existing foreign key matches the explicit numeric naming pattern for that table

File: enginetest/queries/foreign_key_queries.go [hunk 2]
```diff
@@ -2228,6 +2228,188 @@ var ForeignKeyTests = []ScriptTest{
 			},
 		},
 	},
+	{
```
Comment (zachmu) [inline enginetest/queries/foreign_key_queries.go:2231]: Missing test case: create a foreign key called `child5_ibfk_1` on another table (maybe child1), then do

```sql
mysql> alter table child5 add foreign key (j) references theparent (i);
ERROR 1826 (HY000): Duplicate foreign key constraint name 'child5_ibfk_1'
```

So MySQL enforces global uniqueness for all constraint names in a database, across all tables, but will only automatically renumber constraints when they exist on the same table.

File: sql/plan/alter_foreign_key.go [hunk 3]
```diff
@@ -232,6 +233,8 @@ func ResolveForeignKey(ctx *sql.Context, tbl sql.ForeignKeyTable, refTbl sql.For
 		}
 	}
 
+	// TODO: deal with empty foreign key names here
```
Comment (zachmu) [inline sql/plan/alter_foreign_key.go:236]: TODOs seem already taken care of?

File: sql/rowexec/ddl_iters.go [hunk 4]
```diff
@@ -1864,7 +1864,8 @@ func (b *BaseBuilder) executeAlterIndex(ctx *sql.Context, n *plan.AlterIndex) er
 			indexName = strings.Join(n.ColumnNames(), "")
 			if _, ok := indexMap[strings.ToLower(indexName)]; ok {
 				for i := 0; true; i++ {
-					newIndexName := fmt.Sprintf("%s_%d", indexName, i)
+					// MySQL starts at 2 for duplicate indexes
+					newIndexName := fmt.Sprintf("%s_%d", indexName, i+2)
```
Comment (zachmu) [inline sql/rowexec/ddl_iters.go:1868]: Tests are under specified here, need to make sure this behaves correctly in the presence of gaps:

```sql
mysql> create table t (i int);
Query OK, 0 rows affected (0.02 sec)

mysql> alter table t add index i_3(i);
Query OK, 0 rows affected (0.02 sec)
Records: 0  Duplicates: 0  Warnings: 0

mysql> alter table t add index (i);
Query OK, 0 rows affected, 1 warning (0.02 sec)
Records: 0  Duplicates: 0  Warnings: 1

mysql> alter table t add index (i);
Query OK, 0 rows affected, 1 warning (0.02 sec)
Records: 0  Duplicates: 0  Warnings: 1

mysql> alter table t add index (i);
Query OK, 0 rows affected, 1 warning (0.02 sec)
Records: 0  Duplicates: 0  Warnings: 1

mysql> show create table t;
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table
                                                                         |
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `i` int DEFAULT NULL,
  KEY `i_3` (`i`),
  KEY `i` (`i`),
  KEY `i_2` (`i`),
  KEY `i_4` (`i`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci |
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

**Review by zachmu — APPROVED** [review]
Behavior change LGTM but tests are underspecified.

**Pull request merged**